### PR TITLE
add NDArray

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
Add Java NDArray class implementation

### CODE REVIEW

**Comment by lanking520** [comment]
@zachgk 
@piyushghai

File: scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala [hunk 1]
```diff
@@ -1,95 +1,185 @@
...
+class JavaNDArray(val nd : NDArray) {
+
+  def this(arr : Array[Float], shape : JavaShape, ctx : JavaContext) = {
+    this(NDArray.array(arr, shape, ctx))
+  }
+  
+  def serialize() : Array[Byte] = nd.serialize()
+
+  def dispose() : Unit = nd.dispose()
+  def disposeDeps() : NDArray = nd.disposeDepsExcept()
```
Comment (piyushghai) [inline scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala:97]: Should this not be named disposeDepsExcept as well ?

File: scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala [hunk 2]
```diff
@@ -1,95 +1,185 @@
...
-  implicit def toNDArray(jnd: JavaNDArray): NDArray = jnd.ndArray
-
-  val api = NDArray.api
+  implicit def toNDArray(jnd: JavaNDArray): NDArray = jnd.nd
 
   def waitall(): Unit = NDArray.waitall()
 
-  def onehotEncode(indices: JavaNDArray, out: JavaNDArray): JavaNDArray = NDArray.onehotEncode(indices, out)
+  def onehotEncode(indices: JavaNDArray, out: JavaNDArray): JavaNDArray
+  = NDArray.onehotEncode(indices, out)
```
Comment (andrewfayres) [inline scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala]: Don't make changes like this. The line was less than 120 to begin with so there's really not a good reason to split it. On top of that it makes doing PRs more complicated.

File: scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala [hunk 3]
```diff
@@ -1,95 +1,185 @@
...
+  def zeros(ctx: JavaContext, shape: Array[Int]): JavaNDArray = NDArray.zeros(ctx, shape: _*)
+
+  def ones(shape: JavaShape, ctx: JavaContext, dtype: JavaDType.DType): JavaNDArray
+  = NDArray.ones(shape, ctx, dtype)
+
+  def full(shape: JavaShape, value: Float, ctx: JavaContext): JavaNDArray
+  = NDArray.full(shape, value, ctx)
 
   def power(lhs: JavaNDArray, rhs: JavaNDArray): JavaNDArray = NDArray.power(lhs, rhs)
+
```
Comment (andrewfayres) [inline scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala]: There is no write or wrong here, I prefer grouping one liners that are all the same method but putting a break in them is equally valid. However, if you really want to make a lot of style changes it's better to do in a separate PR from functionality changes.

File: scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala [hunk 4]
```diff
@@ -1,95 +1,185 @@
...
   def lesser(lhs: JavaNDArray, rhs: Float): JavaNDArray = NDArray.lesser(lhs, rhs)
 
   def lesserEqual(lhs: JavaNDArray, rhs: JavaNDArray): JavaNDArray = NDArray.lesserEqual(lhs, rhs)
+
   def lesserEqual(lhs: JavaNDArray, rhs: Float): JavaNDArray = NDArray.lesserEqual(lhs, rhs)
 
-  def array(sourceArr: java.util.List[Float], shape: JavaShape, ctx: JavaContext = null): JavaNDArray
-    = NDArray.array(sourceArr.asScala.toArray, shape, ctx)
+  def array(sourceArr: Array[Float], shape: JavaShape, ctx: JavaContext = null): JavaNDArray
```
Comment (andrewfayres) [inline scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala]: I don't think that this is right. A java user will not be making a call to this using Array.

**Review by andrewfayres — CHANGES_REQUESTED** [review]

File: scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala (hunk 2 above)
Comment (lanking520) [reply scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala]: The scalastyle limit the length as 100 instead of 120, so it is breaking scala style

**Comment by lanking520** [comment]
Still the implementation on List will be fairly slow, see the implementation below:
```
  def array(sourceArr: java.util.List[java.lang.Float], shape: JavaShape, ctx: JavaContext = null): JavaNDArray
  = NDArray.array(sourceArr.asScala.map(ele => Float.unbox(ele)).toArray, shape, ctx)
```

File: scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala [hunk 5]
```diff
@@ -1,42 +1,42 @@
...
 
   def waitall(): Unit = NDArray.waitall()
 
   def onehotEncode(indices: JavaNDArray, out: JavaNDArray): JavaNDArray = NDArray.onehotEncode(indices, out)
 
   def empty(shape: JavaShape, ctx: JavaContext, dtype: JavaDType.DType): JavaNDArray = NDArray.empty(shape, ctx, dtype)
-  def empty(shape:  Int *): JavaNDArray = NDArray.empty(shape: _*)
-  def empty(ctx: JavaContext, shape: Int *): NDArray = NDArray.empty(ctx, shape: _*)
-
+  def empty(ctx: JavaContext, shape: Array[Int]): NDArray = NDArray.empty(ctx, shape: _*)
```
Comment (piyushghai) [inline scala-package/core/src/main/scala/org/apache/mxnet/api/java/JavaNDArray.scala]: Should the shape object be an Array ? or can it also be a ArrayList ?

**Review by andrewfayres — APPROVED** [review]
